File: arenasystem-web/src/auth/permissions.js
export const ROLE_LABELS = {
  superadmin_saas: 'Superadmin SaaS',
  dono: 'Dono',
  recepcao: 'Recepcao',
  financeiro: 'Financeiro',
  professor: 'Professor',
  estoque: 'Estoque',
  aluno: 'Aluno',
};

export const ROLE_PERMISSIONS = {
  superadmin_saas: ['*'],
  dono: ['*'],
  administrador: ['dashboard.view', 'alunos.manage', 'reservas.manage', 'pedidos.view', 'turmas.manage', 'checkin.manage', 'access.view', 'access.manage', 'access.override', 'estoque.manage', 'auditoria.view', 'usuarios.view', 'usuarios.manage', 'subscription.view', 'support.view', 'privacy.manage'],
  recepcao: ['dashboard.view', 'alunos.manage', 'reservas.manage', 'pedidos.view', 'turmas.manage', 'checkin.manage', 'access.view', 'access.manage', 'support.view'],
  financeiro: ['dashboard.view', 'mensalidades.manage', 'financeiro.manage', 'subscription.view', 'support.view'],
  professor: ['turmas.own.view', 'checkin.own.manage'],
  estoque: ['estoque.manage', 'pedidos.view'],
  aluno: [],
};

export const ADMIN_MENU_ITEMS = [
  { key: 'saas', permission: 'saas.manage', path: '/admin/saas/arenas', label: 'SaaS', icon: 'settings' },
  { key: 'dashboard', permission: 'dashboard.view', path: '/admin', label: 'Visao Geral', icon: 'overview', exact: true },
  { key: 'solicitacoes', permission: 'alunos.manage', path: '/admin/solicitacoes', label: 'Solicitacoes', icon: 'inbox' },
  { key: 'fila', permission: 'alunos.manage', path: '/admin/fila-espera', label: 'Fila de Espera', icon: 'queue' },
  { key: 'alunos', permission: 'alunos.manage', path: '/admin/alunos', label: 'Alunos', icon: 'users' },
  { key: 'memberplans', permission: 'alunos.manage', path: '/admin/planos-membros', label: 'Planos dos alunos', icon: 'money' },
  { key: 'checkins', permission: 'checkin.manage', path: '/admin/checkins', label: 'Check-ins', icon: 'check' },
  { key: 'mensalidades', permission: 'mensalidades.manage', path: '/admin/mensalidades', label: 'Mensalidades', icon: 'money' },
  { key: 'gestao', permission: 'financeiro.manage', path: '/admin/gestao', label: 'Gestao', icon: 'trend' },
  { key: 'estoque', permission: 'estoque.manage', path: '/admin/estoque', label: 'Estoque', icon: 'stock' },
  { key: 'quadras', permission: 'turmas.manage', path: '/admin/turmas', label: 'Quadras', icon: 'court' },
  { key: 'auditoria', permission: 'auditoria.view', path: '/admin/auditoria', label: 'Auditoria', icon: 'shield' },
  { key: 'subscription', permission: 'subscription.view', path: '/admin/assinatura', label: 'Minha assinatura', icon: 'money' },
  { key: 'team', permission: 'usuarios.view', path: '/admin/equipe', label: 'Equipe', icon: 'users' },
  { key: 'support', permission: 'support.view', path: '/admin/suporte', label: 'Suporte', icon: 'inbox' },
  { key: 'privacy', permission: 'privacy.manage', path: '/admin/privacidade', label: 'Privacidade', icon: 'shield' },
  { key: 'help', permission: 'dashboard.view', path: '/admin/ajuda', label: 'Ajuda', icon: 'inbox' },
  { key: 'saasops', permission: 'saas.manage', path: '/admin/saas/operacao', label: 'Operacao SaaS', icon: 'trend' },
];

export function getRole(user) {
  if (!user) return 'aluno';
  return user.papel_efetivo || user.papel || (user.is_staff || user.tipo === 'admin' ? 'dono' : user.tipo || 'aluno');
}

export function can(user, permission) {
  const role = getRole(user);
  const permissions = ROLE_PERMISSIONS[role] || [];
  return permissions.includes('*') || permissions.includes(permission);
}

export function adminMenuFor(user) {
  return ADMIN_MENU_ITEMS.filter((item) => can(user, item.permission));
}

export function adminNavigationFor(user, { terms, features, hasTenantContext = false } = {}) {
  const vocabulary = terms || {
    pessoa_plural: 'alunos',
    espaco_plural: 'quadras',
  };
  const modules = features || { classes: true, openAccess: false, store: true };

  return adminMenuFor(user)
    .filter((item) => getRole(user) !== 'superadmin_saas' || hasTenantContext || ['saas', 'saasops'].includes(item.key))
    .filter((item) => modules.classes || !['solicitacoes', 'fila', 'quadras'].includes(item.key))
    .filter((item) => modules.classes || modules.openAccess || item.key !== 'checkins')
    .filter((item) => modules.store || item.key !== 'estoque')
    .map((item) => {
      if (item.key === 'alunos') return { ...item, label: capitalize(vocabulary.pessoa_plural) };
      if (item.key === 'memberplans') return { ...item, label: `Planos dos ${vocabulary.pessoa_plural}` };
      if (item.key === 'quadras') return { ...item, label: capitalize(vocabulary.espaco_plural) };
      return item;
    });
}

export function defaultAdminPath(user) {
  if (getRole(user) === 'superadmin_saas') return '/admin/saas/arenas';
  if (can(user, 'dashboard.view')) return '/admin';
  return adminMenuFor(user)[0]?.path || '/';
}

export function hasAdminAccess(user) {
  return adminMenuFor(user).length > 0;
}

export function canUseStudentApp(user) {
  if (!user) return false;
  return user.tipo === 'aluno' || getRole(user) === 'professor' || user.tipo === 'admin' || user.tipo === 'admin_arena' || user.is_staff;
}

export function defaultStudentPath() {
  return '/app';
}

export function initialPathFor(user) {
  if (hasAdminAccess(user)) return defaultAdminPath(user);
  if (canUseStudentApp(user)) return defaultStudentPath();
  return '/acesso-negado';
}

function capitalize(value) {
  return value ? value.charAt(0).toUpperCase() + value.slice(1) : value;
}

File: arenasystem/usuarios/permissions.py
from rest_framework import permissions


PERMISSOES_POR_PAPEL = {
    'superadmin_saas': {'*'},
    'dono': {'*'},
    'administrador': {
        'dashboard.view', 'alunos.view', 'alunos.manage',
        'reservas.view', 'reservas.manage', 'pedidos.view', 'pedidos.manage',
        'checkin.view', 'checkin.manage', 'turmas.view', 'turmas.manage',
        'access.view', 'access.manage', 'access.override',
        'mensalidades.view', 'financeiro.view', 'estoque.view', 'estoque.manage',
        'auditoria.view', 'usuarios.view', 'usuarios.manage', 'subscription.view',
        'support.view', 'support.manage', 'privacy.manage',
    },
    'recepcao': {
        'alunos.view', 'alunos.manage',
        'reservas.view', 'reservas.manage',
        'pedidos.view', 'pedidos.manage',
        'checkin.view', 'checkin.manage',
        'access.view', 'access.manage',
        'turmas.view', 'turmas.manage',
        'dashboard.view',
        'support.view', 'support.manage',
    },
    'financeiro': {
        'mensalidades.view', 'mensalidades.manage',
        'financeiro.view', 'financeiro.manage',
        'reservas.view',
        'dashboard.view',
        'subscription.view',
        'support.view',
    },
    'professor': {
        'turmas.own.view',
        'checkin.own.view', 'checkin.own.manage',
        'alunos.own.view',
    },
    'estoque': {
        'estoque.view', 'estoque.manage',
        'pedidos.view',
    },
    'aluno': {
        'reservas.self.view', 'reservas.self.manage',
        'mensalidades.self.view',
        'pedidos.self.view', 'pedidos.self.manage',
        'turmas.self.view',
        'checkin.self.manage',
        'access.self.view', 'access.self.manage',
    },
}


def papel_usuario(user):
    if not getattr(user, 'is_authenticated', False):
        return None
    return getattr(user, 'papel_efetivo', None) or 'aluno'


def tem_permissao(user, permissao):
    papel = papel_usuario(user)
    permissoes = PERMISSOES_POR_PAPEL.get(papel, set())
    base_allowed = '*' in permissoes or permissao in permissoes
    if papel == 'dono' and permissao in {'subscription.manage', 'ownership.transfer', 'usuarios.manage'}:
        return True
    overrides = getattr(user, 'permission_overrides', None)
    if overrides is None or not getattr(user, 'pk', None):
        return base_allowed
    override = overrides.filter(permission=permissao).values_list('effect', flat=True).first()
    if override == 'deny':
        return False
    if override == 'allow':
        return True
    return base_allowed


class HasArenaPermission(permissions.BasePermission):
    required_permission = None

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        required = getattr(view, 'required_permission', self.required_permission)
        return bool(required and tem_permissao(request.user, required))


class IsSaasSuperAdmin(permissions.BasePermission):
    def has_permission(self, request, view):
        return bool(
            request.user
            and request.user.is_authenticated
            and papel_usuario(request.user) == 'superadmin_saas'
        )


def require_permission(permissao):
    class RequiredPermission(HasArenaPermission):
        required_permission = permissao
    return RequiredPermission
